src/proteus/dswx_hls.py

```diff
@@ -174,12 +239,17 @@
     'evergreen_forest': 201,
 
     # fill value (not masked)
-    'fill_value': 255}
+    'fill_value': UINT8_FILL_VALUE}
 
+'''
+Dict of landcover threshold list:
+  [evergreen, low-intensity developed, high-intensity developed, water/wetland]
+'''
 landcover_threshold_dict = {"standard": [6, 3, 7, 3],
                             "water heavy": [6, 3, 7, 1]}
 
-landcover_nir_threshold = 1200
+MIN_SLOPE_ANGLE = -5
+MAX_SUN_LOCAL_INC_ANGLE = 40
```

Comment:
`MIN_SLOPE_ANGLE` and `MAX_SUN_LOCAL_INC_ANGLE` are default values for the `generate_dswx_layers()` function, which is not used by the DSWx-HLS SAS since the default values for the SAS come from the default runconfig.

I don't want to make `min_slope_angle` and `max_sun_local_inc_angle` required arguments for `generate_dswx_layers()`. Would you have any suggestions on how to improve this?

Reply:
Aye, the input handling makes this a tricky problem.

If we want the user to be able to change it via the runconfig, then perhaps something like was done for the browse image height / width?
https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/dswx_hls.py#L3355-L3358
It's not elegant, but it sets the defaults in a single location, and forces those values to propagate nicely via argument lists.

Alternatively, if these two parameters do not need to be exposed in the runconfig, then they could simply be removed from the .yaml's and from the input parsing, and left as the global constants.

Side note: While pondering this, I noticed that this has already caused a maintainability bug:
In the `_compute_opera_shadow_layer` function, `MIN_SLOPE_ANGLE` is used a default for the parameter `min_slope_angle`. But in the function body, only `MIN_SLOPE_ANGLE` (the global constant) is called, while the actual parameter `min_slope_angle` is never called. This means that any changes to the runconfig's `min_slope_angle` will be silently ignored during that function's execution.

Reply:
Hi @nemo794 , as far as I understood, you are doing the same here:

https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/defaults/dswx_hls.yaml#L94

BTW, maybe we should create constants for default browse images dimensions too:

https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/dswx_hls.py#L3355-L3358

Also, I'd prefer to keep the assignments above in the function call of `generate_dswx_layers()` in:

https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/dswx_hls.py#L2992

Would you mind helping me fix that? I think the bug you mentioned is already fixed, but if not, could you please fix that as well. Thanks!

Reply:
> Hi @nemo794 , as far as I understood, you are doing the same here:
> https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/defaults/dswx_hls.yaml#L94
> 

Yes, a suggested value is included in the default runconfig.yaml file. However, as the code is currently written, if the user supplies any runconfig.yaml file, then the default runconfig.yaml values are completely ignored/overwritten during the call to `_deep_copy()`.

For example, say I'm a user who provides my own runconfig.yaml file, and fills in the input, output, etc. directories but leaves the parameter values blank. When the code is run, the `bin/dswx_hls.py` file will parse the arguments via `parse_runconfig_file()`. However, those values will not be filled in from the default runconfig.yaml file; instead they'll be parsed as (I think?) a Nonetype. (Will need to triple-check the type of the object -- I'm 95% sure it was Nonetype, but it's been a week since I ran these.)

The moral of this is that the .yaml file cannot be trusted to be the source of the defaults, since it gets ignored. We need to have the defaults be set within the `src/proteus/dswx_hls.py` code itself.

> BTW, maybe we should create constants for default browse images dimensions too:
> https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/dswx_hls.py#L3355-L3358
> 

Creating global constants contradicts the goal of having only one source for the these values, from which those values propagate everywhere else in the code. Right now there are two sources: 1) the global constant, and 2) the runconfig.yaml file.

> Also, I'd prefer to keep the assignments above in the function call of generate_dswx_layers() in:
> https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/dswx_hls.py#L2992
> 

At least for the browse images, setting the default values here in the function signature is effectively meaningless when the function is called from the `bin/dswx_hls.py` script. It took a while to figure this out, but what's happening is that when `bin/dswx_hls.py` parses the arguments, the way the parsing function is set up, it will always make sure there is a value supplied for e.g. `args.browse_image_height` -- even if that value is `None`. Then, since `bin/dswx_hls.py` passes that known value of `args.browse_image_height` into `generate_dswx_layers()` for each parameter, that means that the default values supplied in the `generate_dswx_layers()` function signature will never actually be used by the current workflow. This was why I had to do to hacky workaround of [checking for "None"](https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/dswx_hls.py#L3355-L3358).

> Would you mind helping me fix that? I think the bug you mentioned is already fixed, but if not, could you please fix that as well. Thanks!

Ah, seems the version of the code I was looking at was a few hours out of date. You're right, the bug in `_compute_opera_shadow_layer` appears to be fixed now.

Reply:
Default values should come from the default runconfig. These values should be used by DSWx-HLS SAS if the user don't supply his own values, that's why it's a "default runconfig".

It would be great if `generate_dswx_layers()` could use the same values from the runconfig, but I'm not sure how to do that. Please, keep in mind that I'm trying to make `generate_dswx_layers()` callable from code different from the DSWx-HLS SAS so I don't want to make those parameters required. Since, I'm not sure how to do that, at least without an ugly workaround, I think we can use global constants. Notice that these global constants are never used by the DSWx-HLS SAS, as the default values for the SAS come from the default runconfig.

Reply:
100% agree that having all default values come from the default runconfig.yaml would be great, and `generate_dswx_layer()` should definitely have the ability to be called from anywhere -- those intentions have been very apparent from the beginning.

However, that's **not** the current behavior of the existing code and workflow, and the current code is the framework we have to work within. Fixing this properly would likely be a hefty undertaking.

I also understand that these global constants /should/ not be used by the DSWx-HLS SAS, but as mentioned in a previously, during the ongoing process of development, it's too easy for a developer (myself included) to accidentally use these global constants inside a function. When this happened in `_compute_opera_shadow_layer` a couple of days ago, you fortunately caught that bug, but it was caught by chance; relying on chance doesn't seem ideal.

As a quick workaround to set defaults within the existing framework, we could do a series of input checks at the beginning of the `generate_dswx_layers()` function, e.g.:

```
def generate_dswx_layers( ... ):
"""
...
"""
    # Set Defaults
    if min_slope_angle is None:
        min_slope_angle = -5
     
    if max_sun_local_inc_angle is None:
        min_slope_angle = 40

    if hls_thresholds is None:
        hls_thresholds = parse_runconfig_file()

    if scratch_dir is None:
        scratch_dir = '.'
     
     etc.
```

How about something like this?

Reply:
Uhm, the code that you suggested is setting these constants from two places as well (from the runconfig and from inside `generate_dswx_layers()`), so I'm not following the maintainability argument?

I also prefer to make these default values more explicit in the beginning of the module rather than buried inside the code.

Reply:
What I'm suggesting is not ideal. The ideal requires a significant updates to the input handling.

But, what I'm suggesting _does_ force the use of these values to be on a **single critical path** (runconfig first via `bin/dswx_hls.py`, which then must go through the input checks in `src/proteus/dsws_hls.py/generate_dswx_layers()`). Then, if a different function wants access to any one of these values, that function _must_ get the value from `src/proteus/dsws_hls.py/generate_dswx_layers()` by having it passed in via its argument list.

On the other hand, if the values are stored in a global constant, that global constant is **not** on that single critical path. This means that a different function could either get the value via its argument list (which is on the critical path), OR from the global constant. By having these two possible paths for a function to get the value, this is the source of the maintenance issues.

Reply:
They are still defined in two places and values can differ. I still prefer these constants explicit in the beginning of the module and set as default arguments in the function declaration. Much better than buried inside the code. When I find some time, I'll restructure the code so we can move on.

Reply:
This issue was addressed in [e7538c8](https://github.com/nasa/PROTEUS/pull/28/commits/e7538c8f1d936cd04ede6cf026bb51587aa2358d) .